src/pdf_parser/pdf_utils/ocr.py

```
@@ -1,11 +1,329 @@
+):
+    """
+    Combine google layout with layoutparser layout.
+
+    - Replaces layoutparser objects with objects recognised by google API + their text if they overlap sufficiently.
+    - Does not include google objects if they seem to be not part of the main text (e.g. headers, footers, etc.). We
+    use top exclude and bottom exclude to ascertain this.
+
+
+    Args:
+        image: The image the layout is based on.
+        google_layout: List of GoogleTextSegment objects inferred from the structure returned by the Google OCR API.
+        lp_layout: Layout object
+        threshold: Threshold for overlap between google and layoutparser objects. If the overlap is larger than
+            threshold, the layoutparser object is replaced by the google object.
+        top_exclude: Exclude objects from the top of the page if they are above this fraction of the page.
+        bottom_exclude: Exclude objects from the bottom of the page if they are below this fraction of the page.
+
+    Returns:
+        Layout object with google objects replacing layoutparser objects if they overlap sufficiently.
+    """
+    shapely_google = [google_vertex_to_shapely(b.coordinates) for b in google_layout]
+    shapely_layout = [lp_coords_to_shapely_polygon(b.coordinates) for b in lp_layout]
+    # for every block in the google layout, find the fraction of the block that is covered by the layoutparser layout
+    dd_intersection_over_union = defaultdict(list)
+    for ix_goog, google_block in enumerate(shapely_google):
+        for ix_lp, lp_block in enumerate(shapely_layout):
+            # Find the intersection over union of the two blocks.
+            intersection = google_block.intersection(lp_block).area
+            union = google_block.union(lp_block).area
+            dd_intersection_over_union[ix_goog].append(intersection / union)
+
+    # If a google block is covered by a layoutparser block by more than 0.9, we can assume that google OCR has
```

The comment is specific to the threshold that we have set as default so could be misleading. Again I would wrap this in a function.

---

Ok functionalised